File: benchmarks/complex/kmeans.cpp
#include <cstdio>
#include <vector>
using std::vector;
int main(){ int n=50000,kk=16; vector<long> xs(n),ys(n); long seed=987654321;
 for(int i=0;i<n;i++){seed=(seed*1103515245+12345)%2147483648L;xs[i]=seed%1000;seed=(seed*1103515245+12345)%2147483648L;ys[i]=seed%1000;}
 vector<long> cx(kk),cy(kk); for(int i=0;i<kk;i++){cx[i]=xs[i*137%n];cy[i]=ys[i*137%n];}
 vector<long> sumx(kk),sumy(kk),cnt(kk),dist(kk);
 for(int it=0;it<25;it++){ for(int c=0;c<kk;c++){sumx[c]=0;sumy[c]=0;cnt[c]=0;}
  for(int i=0;i<n;i++){ long px=xs[i],py=ys[i];
   for(int c=0;c<kk;c++){long dx=px-cx[c],dy=py-cy[c];dist[c]=dx*dx+dy*dy;}
   int best=0; long bestd=dist[0]; for(int c=1;c<kk;c++) if(dist[c]<bestd){bestd=dist[c];best=c;}
   sumx[best]+=px;sumy[best]+=py;cnt[best]++; }
  for(int c=0;c<kk;c++) if(cnt[c]>0){cx[c]=sumx[c]/cnt[c];cy[c]=sumy[c]/cnt[c];} }
 long cs=0; for(int i=0;i<kk;i++)cs+=cx[i]*31+cy[i]; printf("%ld\n",cs); }

File: benchmarks/complex/kmeans.rs
fn main() {
    let (n, kk) = (50000usize, 16usize);
    let (mut xs, mut ys) = (vec![0i64; n], vec![0i64; n]);
    let mut seed = 987654321i64;
    for i in 0..n { seed=(seed*1103515245+12345)%2147483648; xs[i]=seed%1000; seed=(seed*1103515245+12345)%2147483648; ys[i]=seed%1000; }
    let (mut cx, mut cy) = (vec![0i64; kk], vec![0i64; kk]);
    for i in 0..kk { cx[i]=xs[i*137%n]; cy[i]=ys[i*137%n]; }
    let (mut sumx, mut sumy, mut cnt) = (vec![0i64;kk], vec![0i64;kk], vec![0i64;kk]);
    let mut dist = vec![0i64; kk];
    for _ in 0..25 {
        for c in 0..kk { sumx[c]=0; sumy[c]=0; cnt[c]=0; }
        for i in 0..n {
            let (px,py)=(xs[i],ys[i]);
            for c in 0..kk { let (dx,dy)=(px-cx[c],py-cy[c]); dist[c]=dx*dx+dy*dy; }
            let (mut best,mut bestd)=(0usize,dist[0]);
            for c in 1..kk { if dist[c]<bestd { bestd=dist[c]; best=c; } }
            sumx[best]+=px; sumy[best]+=py; cnt[best]+=1;
        }
        for c in 0..kk { if cnt[c]>0 { cx[c]=sumx[c]/cnt[c]; cy[c]=sumy[c]/cnt[c]; } }
    }
    let mut cs=0i64; for i in 0..kk { cs+=cx[i]*31+cy[i]; } println!("{}",cs);
}
